tests/unit/s2n_build_test.c

```diff
@@ -0,0 +1,148 @@
+* permissions and limitations under the License.
+*/
+#define _GNU_SOURCE
+#include <openssl/crypto.h>
+#include <openssl/opensslv.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+
+#include "s2n_test.h"
+
+/* s2n_test.h redefines OPEN_SSL_VERSION_NUMBER for LIBRESSL */
+#if defined(LIBRESSL_VERSION_NUMBER)
+const long unsigned openssl_version_number = 0x020000000L;
+#else
+const long unsigned openssl_version_number = OPENSSL_VERSION_NUMBER;
+#endif
+
+int main(int argc, char **argv)
+{
+    (void) argc, (void) argv;
+    BEGIN_TEST();
+
+    /*
+    * In the case that libcrypto is dynamicly linked there is a chance that the
+    * build version and the linked version don't match. OPENSSL_VERSION_TEXT is
+    * a macro defined in openssl/crypto.h and should be returned from OpenSSL_version.
+    * In the event that we are dynamicly linked to the wrong libcrypto the function,
+    * call will return a diffrent version text than the one we used at a source level.
+    *
+    * If libcrypto is staticly linked, this is sure to be true.
+    */
+    if (SSLeay() != openssl_version_number) {
+        printf("\nOPENSSL_VERSION_NUMBER == %ld\n", (long int) openssl_version_number);
+        printf("SSLeay() == %ld\n", (long int) SSLeay());
+    }
+    EXPECT_EQUAL(SSLeay(), openssl_version_number);
+
+    /*
+    * The build configurations in CI are defined by S2N_BUILD_PRESET.
+    *
+    */
+    const char *s2n_build_preset = getenv("S2N_BUILD_PRESET");
+    if (s2n_build_preset == NULL) {
+        printf("S2N_BUILD_PRESET is not set.");
+        /* Apparently we aren't in CI. */
+        END_TEST();
+    }
+    printf("S2N_BUILD_PRESET set to %s.", s2n_build_preset);
+
+    /* CMake generated Makefiles can select a default LIBCRYPTO,
+    * lets make sure that didn't happen. */
+    const char *s2n_libcrypto = getenv("S2N_LIBCRYPTO");
+    EXPECT_NOT_NULL(s2n_build_preset);
+
+#define CONTAINS(x) strcasestr(s2n_build_preset, x) != NULL
+#define CHK_LC(x)   EXPECT_EQUAL(strcmp(s2n_libcrypto, x), 0)
+    /* Verify that the environment  */
+    if (CONTAINS("awslc") && CONTAINS("fips")) {
+        CHK_LC("awslc-fips");
+    } else if (CONTAINS("awslc")) {
+        CHK_LC("awslc");
+    } else if (CONTAINS("libressl")) {
+        CHK_LC("libressl");
+    } else if (CONTAINS("boringssl")) {
+        CHK_LC("boringssl");
+    } else if (CONTAINS("openssl") && CONTAINS("fips")) {
+        if (CONTAINS("1-0-2") || CONTAINS("1.0.2")) {
+            CHK_LC("openssl-1.0.2-fips");
+        } else if (CONTAINS("1.1.1") || CONTAINS("1-1-1")) {
+            CHK_LC("openssl-1.1.1-fips");
+        } else if (CONTAINS("3-0") || CONTAINS("3.0")) {
+            CHK_LC("openssl-3.0-fips");
+        } else {
+            printf("\nTest didn't handle this openssl version:\n");
+            printf("S2N_BUILD_PRESET == %s\n", s2n_build_preset);
+            printf("S2N_LIBCRYPTO    == %s\n", s2n_libcrypto);
+            EXPECT_TRUE(0);
```

Comment:
These `EXPECT_TRUE`s should probably be consistent (and preferable not bare integers)

Reply:
How should this code read? I don't quite understand the comment.

Reply:
discussed offline; elsewhere there is `EXPECT_TRUE(false)`

Reply:
Turns out its better to use `FAIL_MSG`.